estimagic/differentiation/numdiff_np.py

```diff
@@ -166,12 +180,162 @@ def _consolidate_one_step_derivatives(candidates, preference_order):
+
+    Args:
+        derivative (np.ndarray): Derivative estimates from Richardson approximation.
+            First axis (axis 0) denotes the potentially multiple estimates. Following
+            dimensions represent the dimension of the derivative, i.e. for a classical
+            gradient ``derivative`` has 2 dimensions, while for a classical jacobian
+            ``derivative`` has 3 dimensions.
+
+        errors (np.ndarray): Error estimates of ``derivative`` estimates. Has the same
+            shape as ``derivative``.
+
+    Returns:
+        - derivative_minimal (np.ndarray): Best derivate estimates chosen with respect
+            to minimizing ``errors``. Note that the best values are selected
+            element-wise. Has shape ``(derivative.shape[1], derivative.shape[2])``.
+
+        - error_minimal (np.ndarray): Minimal errors selected element-wise along axis
+            0 of ``errors``.
+
+    """
+    if derivative.shape[0] == 1:
+        derivative_minimal = np.squeeze(derivative, axis=0)
+        error_minimal = np.squeeze(errors, axis=0)
+    else:
+        try:
+            minimizer = np.nanargmin(errors, axis=0)
+
+            derivative_minimal = np.take_along_axis(
+                derivative, minimizer[np.newaxis, :], axis=0
+            )
+            derivative_minimal = np.squeeze(derivative_minimal, axis=0)
+            error_minimal = np.nanmin(errors, axis=0)
+        except ValueError:
+            raise ValueError(
+                "Richardson extrapolation failing --cannot compute minimum of errors "
+                "along axis 0. "
+            )
+
+    return derivative_minimal, error_minimal
+
+
+def _get_best_estimate_along_methods(derivatives, errors):
+    """Extract best derivative estimate over different methods.
+
+    Given that for each method, where one method can be for example central differences
+    with two num_terms (see above), we have selected a single best derivative estimate,
+    we select the best derivative estimates element-wise over different methods, where
+    again best is defined as minimizing the approximation error.
+
+    Args:
+        derivatives (OrderedDict): Dictionary containing derivative estimates for
+            different methods.
+
+        errors (OrderedDict): Dictionary containing error estimates for derivates stored
+            in ``derivatives``.
+
+    Returns:
+        jac_minimal (np.ndarray): The optimal derivative estimate over different
+            methods.
+
+    """
+    errors = np.stack(list(errors.values()))
+    derivatives = np.stack(list(derivatives.values()))
+
+    if derivatives.shape[0] == 1:
+        jac_minimal = np.squeeze(derivatives, axis=0)
+    else:
+        try:
+            minimizer = np.nanargmin(errors, axis=0)
+
+            jac_minimal = np.take_along_axis(
+                derivatives, minimizer[np.newaxis, :], axis=0
+            )
+            jac_minimal = np.squeeze(jac_minimal, axis=0)
+        except ValueError:
```

Comment:
Why only except ValueErrors? Are they frequent here? What causes them?

Reply:
The ``ValueError`` can be caused if there are all-NaN slices in the errors, when computing the nanargmin, see https://numpy.org/doc/1.18/reference/generated/numpy.nanargmin.html.

I don;t think this occurs often.

I'm not sure anymore if we need a try-except block, it seems less clean / informative than just letting np.nanargmin raise the error.

Reply:
Yes, the nanargmin error message is good enough.